### PR TITLE
Implemented DirAccess get_drive and get_drive_count for Linux

### PR DESCRIPTION
**Pull request by marcelofg55**
File dialog will now show the volumes on Linux.

### CODE REVIEW

**Comment by hpvb** [comment]
I don't think Linux users will expect all of their mounts to show up like that. If we want to provide a list of 'drives' for Linux I instead suggest showing xdg user dirs, and perhaps mounted devices that are not in fstab. This would for instance show people's separate /var partitions, people's tmpfs on /tmp and other such things. 

On my system with the current logic the following devices would be shown:

* /tmp
* /boot
* /boot/efi
* /var/lib/nfs/rpc_pipefs

What would not be show would be USB sticks plugged in by me, as on modern desktops they show up as /run/user/<uid>/media/... which are filtered.

If you want to get really fancy you could only show mounted devices mounted by the user, Gnome and KDE have APIs for that. If a user's not using either of those maybe fall back to showing mounted devices not in /etc/fstab. 

Regardless I suggest looking at what Nautilus and KDE Filemanagers do with external devices.

**Comment by hpvb** [comment]
I have another suggestion for an algorithm maybe:

* Read all mounted devices from /proc/mounts (or /etc/mtab, same thing though)
* Remove all mountpoints mentioned in /etc/fstab
* Remove all mountpoints where the device does not start with '/dev'
* Add the directory pointed to by the ${HOME} environment
* Add directories in ~/.config/gtk-3.0/bookmarks (that start with file:///)

**Comment by marcelofg55** [comment]
Oh, I see, I've only tested this with Ubuntu and the current filter logic worked fine with it. I'll keep working out the filter logic, thanks for the suggestions :).
Btw, what distro do you use?

**Comment by hpvb** [comment]
I use Fedora 26, but you really can't use any distro specific things here. Installers allow for a large amount of variation in the partitioning of a system. Relying purely on the mounted devices is never going to produce a result that's helpful for a Linux user.

**Comment by Noshyaar** [comment]
Discussed in the IRC (http://godot.eska.me/irc-logs/devel/2017-09-11.log from 23:26)

```
<reduz> yeah i am mixed about that one, Linux does not have drive counts but I guess you can access mounts
<reduz> so not really sure if there is a standard way to get a list of mount points
[...]
<reduz> i dont know either, but i guess we can ask the author to do more research
```

**Comment by marcelofg55** [comment]
I managed to do pretty much what was suggested by @hpvb:
virtualbox_ubuntu_11_09_2017_19_04_25

@Noshyaar @reduz Is that okay for you guys? I was not sure if including $HOME and the bookmarks from ~/.config/gtk-3.0/bookmarks would be intended but I guess it's handy on the file dialog to have those.

**Comment by Noshyaar** [comment]
The list seems fine to me. (Not sure about Music, Pictures and Video folders.)

**Comment by marcelofg55** [comment]
Documents, Downloads, Music, Pictures and Videos come from parsing ~/.config/gtk-3.0/bookmarks, it will include whatever users have in the bookmarks file, but it's all or nothing :).

**Comment by hpvb** [comment]
My opinion is that if we're going to do this at all this is probably what it should look like.

This is the standard GTK file dialog:
screenshot from 2017-09-12 03-25-26

KDE has a similar system, but uses a different 'favorites' file. 

If we don't want to explicitly support the 'favorites' or 'places' like this then we can just drop that. I think however that for most Linux users this is the most helpful analogy to 'drives'. If we really only want to support 'drives' then I suggest only adding '/', '${HOME}', and mounted devices starting with '/dev' not in fstab.

At any rate I like this current implementation! 👍

**Comment by marcelofg55** [comment]
@hpvb Check the updated commit, it's pretty much as you suggested but I didn't filter devices in /etc/fstab though, wouldn't that remove partitions that users mount permanently? I don't see a reason why we want to filter those.

**Comment by hpvb** [comment]
@marcelofg55 that text was primarily intended to explain why I suggested the change at all. I did end the post with 'I like what you did' :)

About the fstab thing: most of the time permanently mounted devices aren't removable and aren't useful devices to show from a user application. For instance showing most Linux users their '/boot' partition as a drive in Godot isn't going to really help anyone.  And on EFI systems yo would see /boot and /boot/efi as 'drives' too.

Have a look at the locations shown in the gnome file dialog: you don't see my '/boot' and '/boot/efi' drives there. Others may have installed their systems with a separate /var or /var/log partition. These mountpoints are useless for any 'regular use' of the system. This is also where the gtk bookmarks come in.

If you do use multiple mountpoints to store user-writable data permanently most people will have added them to those bookmarks already to make them easily accessible. If this is still undesirable then I suggest we make a longer list of 'implementation detail' mountpoints that should be filtered if you do want to show some permanently mounted devices. A list like that would look something like:

* anything starting with /boot, /tmp, /etc, /var
* /usr, /usr/local, /opt, /srv
(I'm probably missing some)

or maybe more simply: Only show permanent mounts in /mnt and /media, or that start with /home. 

If we must have some permanently mounted devices shown I'd prefer the last option. That would make it look something like this:

```c++
static bool _filter_drive(struct mntent *mnt) {
	if (strncmp(mnt->mnt_fsname, "/dev", 4) != 0) {
		return false;
	}

	if ((strncmp(mnt->mnt_dir, "/media", 6) == 0)) ||
		(strncmp(mnt->mnt_dir, "/mnt", 4) == 0) ||
		(strncmp(mnt->mnt_dir, "/home", 5)) {
		return true;
	}

	return false;
}
```

Some care will have to be taken that no duplicates get added this way though, particularly for ${HOME}

**Comment by marcelofg55** [comment]
@hpvb Yeah, you're right. I'm going to go with that _filter_drive code you did, seems a good way to keep it simple and make it work as Linux users would expect. Thanks for the detailed explanation :).

**Comment by hpvb** [comment]
I forgot an == 0 in my example code 🥇 In case you were going to copy/paste that code. I fixed the comment now.

**Comment by marcelofg55** [comment]
@hpvb Don't worry, I saw that.

I've updated the commit and this should be ready for merging.

**Comment by hpvb** [comment]
This looks great! Small nitpick though: Can you split the changes to the SConstruct file out into a different commit and explain what it's for?

**Comment by marcelofg55** [comment]
Yeah, sure. It's for checking for mntent.h basically, else this will fail to compile on Mac for example. I'll try to split the commit.

**Comment by hpvb** [comment]
Great, I'll merge this once Travis is happy.

**Review by hpvb — APPROVED** [review]

File: drivers/unix/dir_access_unix.cpp [hunk 1]
```diff
@@ -176,13 +183,77 @@ void DirAccessUnix::list_dir_end() {
 	_cisdir = false;
 }
 
+#ifdef HAVE_MNTENT
+static bool _filter_drive(struct mntent *mnt) {
+	if (strncmp(mnt->mnt_fsname, "/dev", 4) == 0) {
```
[the remark this answers is not in the document]
Comment (akien-mga) [reply drivers/unix/dir_access_unix.cpp]: Well I have devices in `/etc/fstab` that I definitely would want to see in my list of "drives":
```
# Entry for /dev/sda9 :
UUID=5d469a85-d683-4d71-812c-6e67b704104e / ext4 noatime,acl 1 1
# Entry for /dev/sda7 :
UUID=ae7134a1-78bf-4b7b-9080-679b3e2079fa /home ext4 noatime,acl 1 2
# Entry for /dev/sda6 :
UUID=ece54d32-fbd5-4c86-9503-a568ba1d4f3d /media/data ext4 noatime,acl 1 2
# Entry for /dev/sda1 :
UUID=540b82d6-74e1-4948-b32d-efa270a5f3f5 /media/data2 ext4 defaults 1 2
# Entry for /dev/sda8 :
UUID=c6548451-c66d-40dd-9447-da7f4218f3bb /media/testinst ext4 noatime,acl 1 2
# Entry for /dev/sda5 :
UUID=5f8a8879-2c41-4888-a176-eb1435bcff94 /media/testinstall ext4 noatime,acl 1 2
```

[the remark this answers is not in the document]
Comment (hpvb) [reply drivers/unix/dir_access_unix.cpp]: That's not implemented in this PR now. Also, what matters is what they start with in /etc/mtab. There will be /dev entries there:

```
/etc/fstab: UUID=ddccd5a4-017e-4bf2-a587-002587105e20 /                       xfs     defaults        0 0
/etc/mtab: /dev/sda4 / xfs rw,seclabel,relatime,attr2,inode64,noquota 0 0
```

The current implementation will show you all the drives you listed in this comment.

**Pull request merged**